File: problem1.cpp
// Time Complexity : O(n^2)
// Space Complexity : O(n^2)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> allpaths;
        vector<int> currpath;
        findPaths(root, targetSum, currpath, allpaths);
        return allpaths;  
    }

    void findPaths(TreeNode* root, int targetSum, vector<int>& currpath, vector<vector<int> >& allpaths) {
        if(!root)
            return;
        currpath.push_back(root->val);

        targetSum = targetSum - root->val;

        if(targetSum==0 && !root->left && !root->right)
            allpaths.push_back(currpath);

        findPaths(root->left,targetSum, currpath, allpaths);
        findPaths(root->right,targetSum, currpath, allpaths);
        
        currpath.pop_back();

        }
};

File: Problem1.cs
public class Solution
    {
        // Time Complexity : O(n) - traverse all nodes
        // Space Complexity : O(h) - h is height of the tree- recursion stack will at the max have h function calls
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

        IList<IList<int>> result;
        public IList<IList<int>> PathSum(TreeNode root, int targetSum)
        {
            result = new List<IList<int>>();
            helper(root, targetSum, 0, new List<int>());
            return result;
        }

        public void helper(TreeNode root, int targetSum, int currSum, List<int> path)
        {
            //Base case
            if (root == null) return;

            currSum = currSum + root.val;
            //action
            path.Add(root.val);
            //Console.WriteLine(root.val);
            if (root.left == null && root.right == null)
            {
                if (currSum == targetSum)
                {
                    List<int> resPath = new List<int>();
                    for (int i = 0; i < path.Count; i++)
                    {
                        resPath.Add(path[i]);
                        Console.WriteLine(path[i]);
                    }
                    Console.WriteLine();
                    result.Add(resPath);
                }
            }

            //recurse
            helper(root.left, targetSum, currSum, path);

            helper(root.right, targetSum, currSum, path);

            //backtrack
            path.RemoveAt(path.Count - 1);

        }
    }
